lib/IDE/CodeCompletion.cpp

```diff
@@ -2606,10 +2606,18 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
       assert(isa<ConstructorDecl>(CurrDeclContext) &&
              "can call super.init only inside a constructor");
       needInit = true;
-    } else if (addName.empty() && HaveDot &&
-               Reason == DeclVisibilityKind::MemberOfCurrentNominal) {
-      // This case is querying the init function as member
-      needInit = true;
+    } else if (addName.empty()) {
+      auto ReasonIsValid =
+        Reason == DeclVisibilityKind::MemberOfCurrentNominal ||
+        Reason == DeclVisibilityKind::MemberOfSuper ||
+        Reason == DeclVisibilityKind::MemberOfProtocolImplementedByCurrentNominal;
+      auto instTy = ExprType->castTo<AnyMetatypeType>()->getInstanceType();
```

Comment:
I don't like the name `isOnMetatype` at all. We don't call initializers on metatypes to begin with, right? I renamed it to `isOnType` and changed the code to not have a need to use `ExprType` and here at all. I decided not to touch anything else **yet**, since `getTypeOfMember`, which is the only place where `baseType` is used,  is incorrect as well (see [SR-7802](https://bugs.swift.org/browse/SR-7802)). after fixing that, we might not need `baseType` anymore and the problem gets fixed by itself.